streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -2,18 +2,40 @@ package zio.stream.experimental
 
 import zio._
 
-class ZStream[-R, +E, -I, +O, +A](val process: ZManaged[R, Nothing, ZStream.Pull[R, E, I, O, A]]) { self =>
-  def map[B](f: A => B): ZStream[R, E, I, O, B] =
-    ZStream(self.process.map(pull => i => pull(i).map(f)))
+class ZStream[-R, +E, -I, +B, +A](val process: ZManaged[R, Nothing, ZStream.Control[R, E, I, B, A]]) { self =>
+  import ZStream.Control
+
+  def map[C](f: A => C): ZStream[R, E, I, B, C] =
+    ZStream {
+      self.process.map { control =>
+        Control(
+          control.pull.map(f),
+          control.command
+        )
+      }
+    }
 }
 
 object ZStream {
-  type Pull[-R, +E, -I, +O, +A] = I => ZIO[R, Either[O, E], A]
+
+  final case class Control[-R, +E, -I, +B, +A](
+    pull: Pull[R, E, B, A],
+    command: Command[R, E, I]
+  )
+
+  type Pull[-R, +E, +B, +A] = ZIO[R, Either[E, B], A]
+
+  type Command[-R, +E, -I] = I => ZIO[R, E, Any]
```

Comment:
I am not sure this type alias adds much (just adds a level of indirection without saving much typing). I think I'd delete it. Or if not delete it, then add the corresponding `type Query` for `ZSink`.